### ISSUE TITLE
KEYRING_PROPERTY_SSH_KEY_FINGERPRINT doesn't work with multiple identities in ssh-add

### ISSUE THREAD

**Issue by hans2520** [p-sherratt/sagecipher#1]
Python 3.6.9
keyring Version: 22.0.1
sagecipher, version 0.7.5

```
keyring set ansible-vault ${USER} < <(echo "super")
Please select from the following keys...
[1] ssh-rsa ...
[2] ssh-rsa ...
Selection (1..2): Traceback (most recent call last):
  File "/home/ubuntu/.local/bin/keyring", line 8, in <module>
    sys.exit(main())
  File "/home/ubuntu/.local/lib/python3.6/site-packages/keyring/cli.py", line 135, in main
    return cli.run(argv)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/keyring/cli.py", line 68, in run
    return method()
  File "/home/ubuntu/.local/lib/python3.6/site-packages/keyring/cli.py", line 85, in do_set
    set_password(self.service, self.username, password)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/keyring/core.py", line 60, in set_password
    get_keyring().set_password(service_name, username, password)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/keyring/backends/chainer.py", line 58, in set_password
    return keyring.set_password(service, username, password)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/keyrings/alt/file_base.py", line 123, in set_password
    password_encrypted = self.encrypt(password.encode('utf-8'), assoc)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/sagecipher/keyring.py", line 33, in encrypt
    fingerprint = self.ssh_key_fingerprint
  File "/home/ubuntu/.local/lib/python3.6/site-packages/sagecipher/keyring.py", line 24, in ssh_key_fingerprint
    return prompt_for_key()
  File "/home/ubuntu/.local/lib/python3.6/site-packages/sagecipher/cipher.py", line 125, in prompt_for_key
    i = int(input("Selection (1..%s): " % len(keys)))
EOFError: EOF when reading a line
```

Looking at keyring.py and debgging at line 22, we see the following:
```
(Pdb++) pprint(getmembers(self))
[('__abstractmethods__', frozenset()),
 ('__class__', <class 'sagecipher.keyring.Keyring'>),
 ('__delattr__',
  <method-wrapper '__delattr__' of Keyring object at 0x7fe1163a67b8>),
 ('__dict__', {}),
 ('__dir__', <built-in method __dir__ of Keyring object at 0x7fe1163a67b8>),
 ('__doc__', None),
 ('__eq__', <method-wrapper '__eq__' of Keyring object at 0x7fe1163a67b8>),
 ('__format__',
  <built-in method __format__ of Keyring object at 0x7fe1163a67b8>),
 ('__ge__', <method-wrapper '__ge__' of Keyring object at 0x7fe1163a67b8>),
 ('__getattribute__',
  <method-wrapper '__getattribute__' of Keyring object at 0x7fe1163a67b8>),
 ('__gt__', <method-wrapper '__gt__' of Keyring object at 0x7fe1163a67b8>),
 ('__hash__', <method-wrapper '__hash__' of Keyring object at 0x7fe1163a67b8>),
 ('__init__', <method-wrapper '__init__' of Keyring object at 0x7fe1163a67b8>),
 ('__init_subclass__',
  <built-in method __init_subclass__ of KeyringBackendMeta object at 0x25ade38>),
 ('__le__', <method-wrapper '__le__' of Keyring object at 0x7fe1163a67b8>),
 ('__lt__', <method-wrapper '__lt__' of Keyring object at 0x7fe1163a67b8>),
 ('__module__', 'sagecipher.keyring'),
 ('__ne__', <method-wrapper '__ne__' of Keyring object at 0x7fe1163a67b8>),
 ('__new__', <built-in method __new__ of type object at 0x9d17a0>),
 ('__reduce__',
  <built-in method __reduce__ of Keyring object at 0x7fe1163a67b8>),
 ('__reduce_ex__',
  <built-in method __reduce_ex__ of Keyring object at 0x7fe1163a67b8>),
 ('__repr__',
  <bound method FileBacked.__repr__ of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('__setattr__',
  <bound method Keyring.__setattr__ of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('__sizeof__',
  <built-in method __sizeof__ of Keyring object at 0x7fe1163a67b8>),
 ('__str__',
  <bound method KeyringBackend.__str__ of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('__subclasshook__',
  <built-in method __subclasshook__ of KeyringBackendMeta object at 0x25ade38>),
 ('__weakref__', None),
 ('_abc_cache', <_weakrefset.WeakSet object at 0x7fe1163a6518>),
 ('_abc_negative_cache', <_weakrefset.WeakSet object at 0x7fe1163a6588>),
 ('_abc_negative_cache_version', 203),
 ('_abc_registry', <_weakrefset.WeakSet object at 0x7fe1163a64e0>),
 ('_classes',
  {<class 'keyring.backends.fail.Keyring'>,
   <class 'keyrings.alt.Gnome.Keyring'>,
   <class 'keyrings.alt.Google.KeyczarDocsKeyring'>,
   <class 'keyrings.alt.Google.DocsKeyring'>,
   <class 'keyrings.alt.Windows.RegistryKeyring'>,
   <class 'keyrings.alt.Windows.EncryptedKeyring'>,
   <class 'keyrings.alt.file.PlaintextKeyring'>,
   <class 'keyrings.alt.file.EncryptedKeyring'>,
   <class 'keyrings.alt.multi.MultipartKeyringWrapper'>,
   <class 'keyrings.alt.pyfs.BasicKeyring'>,
   <class 'keyrings.alt.pyfs.PlaintextKeyring'>,
   <class 'keyrings.alt.pyfs.EncryptedKeyring'>,
   <class 'keyrings.alt.pyfs.KeyczarKeyring'>,
   <class 'keyring.backends.kwallet.DBusKeyring'>,
   <class 'keyring.backends.kwallet.DBusKeyringKWallet4'>,
   <class 'keyring.backends.SecretService.Keyring'>,
   <class 'keyring.backends.Windows.WinVaultKeyring'>,
   <class 'keyring.backends.chainer.ChainerBackend'>,
   <class 'keyring.backends.macOS.Keyring'>,
   <class 'sagecipher.keyring.Keyring'>}),
 ('_ensure_file_path',
  <bound method Keyring._ensure_file_path of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('_generate_assoc',
  <bound method Keyring._generate_assoc of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('_write_config_value',
  <bound method Keyring._write_config_value of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('decrypt',
  <bound method Keyring.decrypt of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('delete_password',
  <bound method Keyring.delete_password of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('encrypt',
  <bound method Keyring.encrypt of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('file_path', '/home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg'),
 ('file_version', None),
 ('filename', 'sagecipher_pass.cfg'),
 ('get_credential',
  <bound method KeyringBackend.get_credential of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('get_password',
  <bound method Keyring.get_password of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('get_viable_backends',
  <bound method KeyringBackend.get_viable_backends of <class 'sagecipher.keyring.Keyring'>>),
 ('name', 'keyring Keyring'),
 ('priority', 1),
 ('scheme', '[PBKDF2] AES256.CBC (sagecipher)'),
 ('set_password',
  <bound method Keyring.set_password of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('set_properties_from_env',
  <bound method KeyringBackend.set_properties_from_env of <Keyring with [PBKDF2] AES256.CBC (sagecipher) v.1.0 at /home/ubuntu/.local/share/python_keyring/sagecipher_pass.cfg>>),
 ('version', '1.0'),
 ('viable', True)]
```

Inserting the statement on `self.set_properties_from_env()` line 22 fixes this issue. We then see the following:
```
(Pdb++) pprint(getmembers(self))
...
 ('ssh_key_fingerprint', '...'),
 ...]
```
It seems set_properties_from_env was added in keyring 19.3.0 -- https://pypi.org/project/keyring/19.3.0/

**Comment by hans2520** [comment]
Here's a format-patch for convenience
[0001-Fixing-KEYRING_PROPERTY_SSH_KEY_FINGERPRINT.patch.txt](https://github.com/p-sherratt/sagecipher/files/5979956/0001-Fixing-KEYRING_PROPERTY_SSH_KEY_FINGERPRINT.patch.txt)

**Comment by p-sherratt** [comment]
this is odd - `set_properties_from_env` should be invoked by keyring itself when the backend is loaded: https://github.com/jaraco/keyring/blob/v22.0.1/keyring/core.py#L26

are you sure that your `KEYRING_PROPERTY_`... was exported from your shell to the environment when you first tested?

or are you trying to consume this directly instead of using `keyring.get_keyring()` ?  in which case you would need to call `set_properties_from_env()` yourself after instantiating `sagecipher.Keyring`.

**Comment by gzm55** [comment]
i test the latest keyring+sagecipher, but can not reproduce the issue from the cli command.

**Comment by hans2520** [comment]
@p-sherratt 

This is the source code reproducing the issue:

```
    ...
    export KEYRING_PROPERTY_SSH_KEY_FINGERPRINT=$(print_md5_from_file ${*})
    keyring set ${key} ${USER} < <(echo "${VAULT_PASS}")
```

With the patch provided above applied, everything works; without it, we find there are no env properties set within the sagecipher library when it hits the applicable point in execution.

In the code you provided, this occurs within the method `set_keyring` -- that method or the keyring core doesn't appear anywhere in the stack traces I provided in original post -- are we certain it will always be called?

On a system where this happens (so far all of them), this is what is shown for `--list-backends`

```
ubuntu$ keyring --list-backends
sagecipher.keyring.Keyring (priority: 1)
keyring.backends.chainer.ChainerBackend (priority: 10)
keyrings.alt.file.PlaintextKeyring (priority: 0.5)
keyring.backends.fail.Keyring (priority: 0)
```

After installing keyring via pip, this is the default backend priority list. Do things not happen in the same order if ChainerBackend is invoked first? I notice that on other systems not using sagecipher (e.g. Mac OS -- macOS.Keyring), that they will by default ranked with higher priority than ChainerBackend.

**Comment by p-sherratt** [comment]
I believe the issue here is in the `ChainerBackend` :  set_properties_from_env() is called against the chainer after it is selected as highest priority backend by keyring.core.get_keyring(), but then as each viable/subordinate backend is instantiated for the chainer, its own set_properties_from_env() method is _not_ called.

This really needs to be fixed in keyring itself so I'll raise a bug report there.

If you have set this backend explicitly in your environment, i.e.
```
export PYTHON_KEYRING_BACKEND=sagecipher.keyring.Keyring
```
..then I think you won't have the problem?

**Comment by hans2520** [comment]
Right now, our use case will rarely have multiple key identities in the same ssh-agent, as we've taken steps to ensure that we will use a new ssh-agent if there is one loaded in the environment at script-execution time. So I'll take your word for it for now and add in the export just to be extra safe. I put a comment in the kerying issue https://github.com/jaraco/keyring/issues/495, though, I would think the chainer backend should set the environments on chained backends that are invoked.

> i test the latest keyring+sagecipher, but can not reproduce the issue from the cli command.

@gzm55 , you would need to ensure that you have ChainerBackend with higher priority than sagecipher, and that you have added multiple key identities to your SSH agent so that sagecipher will prompt to tty for a choice as to which identity to select. If these conditions are not met, then the issue will not be reproduced.

**Comment by gzm55** [comment]
@hans2520 I forced the backend by setting PYTHON_KEYRING_BACKEND to sagecipher, therefore will not trigger the issue.

**Comment by p-sherratt** [comment]
fixed in keyring v23.0.0
